snap/snapshots_test.go

```diff
@@ -48,6 +48,109 @@ type snapshotSuite struct{}
+		"must-start-with-2": {snap.SnapshotOptions{Exclude: []string{"$SNAP_COMMON_STUFF"}}, mustStartWithError},
+		"path-not-clean-1":  {snap.SnapshotOptions{Exclude: []string{"$SNAP_DATA/../../meh"}}, pathNotCleanError},
+		"path-not-clean-2":  {snap.SnapshotOptions{Exclude: []string{"$SNAP_DATA/"}}, pathNotCleanError},
+		"invalid-chars-1":   {snap.SnapshotOptions{Exclude: []string{"$SNAP_DATA/{one,two}"}}, pathInvalidCharsError},
+		"invalid-chars-2":   {snap.SnapshotOptions{Exclude: []string{"$SNAP_DATA/tree**"}}, pathInvalidCharsError},
+		"invalid-chars-3":   {snap.SnapshotOptions{Exclude: []string{"$SNAP_DATA/foo[12]"}}, pathInvalidCharsError},
+		"invalid-chars-4":   {snap.SnapshotOptions{Exclude: []string{"$SNAP_DATA/bar?"}}, pathInvalidCharsError},
+	}
+
+	for name, test := range testMap {
+		snapshotOptionsCopy := test.snapshotOptions
+		c.Check(test.snapshotOptions.Validate(), ErrorMatches, test.expectedError, Commentf("test: %q", name))
+		c.Check(test.snapshotOptions, DeepEquals, snapshotOptionsCopy)
+	}
+}
+
+func (s *snapshotSuite) TestValidateHappy(c *C) {
+	testMap := map[string]struct {
+		snapshotOptions snap.SnapshotOptions
+	}{
+		"exclude-empty":   {snap.SnapshotOptions{Exclude: []string{}}},
+		"exclude-typical": {snap.SnapshotOptions{Exclude: snapshotHappyExpectedExclude}},
+	}
+
+	for name, test := range testMap {
+		c.Check(test.snapshotOptions.Validate(), IsNil, Commentf("test: %q", name))
+	}
+}
+
+func (s *snapshotSuite) TestMergeDynamicExcludesError(c *C) {
+	snapshotOptions := snap.SnapshotOptions{Exclude: snapshotHappyExpectedExclude}
+	dynamicExcludes := []string{"/home/ubuntu"}
+	snapshotOptionsCopy := snapshotOptions
+	c.Check(snapshotOptions.MergeDynamicExcludes(dynamicExcludes), ErrorMatches, "snapshot exclude path must start with one of.*")
+	c.Check(snapshotOptions, DeepEquals, snapshotOptionsCopy)
+}
+
+func (s *snapshotSuite) TestMergeDynamicExcludesHappy(c *C) {
+	snapshotOptions := snap.SnapshotOptions{Exclude: snapshotHappyExpectedExclude}
+	snapshotOptionsMerged := snap.SnapshotOptions{
+		Exclude: append(snapshotHappyExpectedExclude, snapshotHappyExpectedExclude...),
+	}
+
+	testMap := map[string]struct {
+		dynamicExcludes []string
+		expectedOptions snap.SnapshotOptions
+	}{
+		"exclude-nil":     {nil, snapshotOptions},
+		"exclude-empty":   {[]string{}, snapshotOptions},
+		"exclude-typical": {snapshotHappyExpectedExclude, snapshotOptionsMerged},
+	}
+
+	for name, test := range testMap {
+		snapshotOptionsCopy := snapshotOptions
+		c.Check(snapshotOptionsCopy.MergeDynamicExcludes(test.dynamicExcludes), IsNil, Commentf("test: %q", name))
+		c.Check(snapshotOptionsCopy, DeepEquals, test.expectedOptions, Commentf("test: %q", name))
+	}
+}
+
+func (s *snapshotSuite) TestUnset(c *C) {
+	testMap := map[string]struct {
+		options   *snap.SnapshotOptions
+		isUnset   bool
+		mustPanic bool
+	}{
+		"options-nil":     {options: nil, mustPanic: true},
```

Comment:
This is just testing that dereferencing a nil pointer panics? Or at most that the method uses a field. It doesn't seem necessary

Reply:
Removed